### ISSUE TITLE
be able to depend on a configuration of a target

### ISSUE THREAD

**Issue by dmitry-stepanenko** [nrwl/nx#15064]
- [x] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description

Right now dependsOn syntax allows to specify only the name of the target. However, there're certain scenarios where it would be beneficial to have an ability to depend on a certain configuration of this target
## Motivation

Consider the following scenario (it is a bit simplified just to get an idea): 
- there's a build target, that should use a certain config based on the configuration
- there're 2 build targets, that use different executors, one of them uses custom logic to run a deployment script, another uses a certain 3rd-party executor

With this setup it is not possible to properly specify the dependsOn, because for `deploy.custom` I need to depend on `myapp:build` and for `deploy.cloudflare` I need to depend on `myapp:build:cloudflare`. Of course, there're several workarounds: 

- I can define an empty configuration `cloudflare` for the `deploy.cloudflare` target, but it will only work if command is run as `nx run myapp:deploy.cloudflare:cloudflare`, because even a default configuration is not inherited by dependant tasks. 
- I can define an intermediate target and depend on it (this is demonstrated in the config sample below with a separate `build-cloudflare` target.

```json
{
    "name": "myapp",
    "projectType": "application",
    "sourceRoot": "apps/myapp/src",
    "targets": {
      "build": {
        "executor": "my-executor",
        "options": {
          "configFile": "apps/myapp/config.ts"
        },
        "configurations": {
          "cloudflare": {
            "configFile": "apps/myapp/config.cloudflare.ts"
          }
        }
      },
      "deploy.custom": {
        "executor": "nx:run-commands",
        "options": {
          "command": "node scripts/deploy.js"
        },
        "dependsOn": ["build"]
      },
      "deploy.cloudflare": {
        "executor": "@k11r/nx-cloudflare-wrangler:deploy-page",
        "options": {
          "branch": "master",
          "projectName": "myapp",
          "dist": "dist/apps/myapp/client"
        },
        "dependsOn": ["build-cloudflare"]
      },
      "build-cloudflare": {
        "executor": "nx:run-commands",
        "options": {
          "command": "npx nx run myapp:build:cloudflare"
        }
      }
    }
  }
 ```
There're some additional arguments mentioned in this discussion https://github.com/nrwl/nx/discussions/9236

## Suggested Implementation

An ideal scenario would be to specify a certain configuration explicitly in the dependsOn
```json
      "deploy.cloudflare": {
        "executor": "@k11r/nx-cloudflare-wrangler:deploy-page",
        "options": {
          "branch": "master",
          "projectName": "myapp",
          "dist": "dist/apps/myapp/client"
        },
        "dependsOn": ["build:cloudflare"]
      },
```

I have even started implementing this, but reached a point where I guess it needs to be discussed. 

The way I'm expecting this to work is:
- If target configuration is explicitly specified, it should be applied to this target and all descendants
- If any further descendant target has another "dependsOn" with explicit configuration, this explicit configuration should override the parent one

The problem is that this approach creates certain ambiguity, because there's a chance that certain target is specified as a dependency of multiple other targets with different configurations. So that dependency map may look like this
```
{
  'app1:build': ['lib1:build:production', 'lib3:build:production'],
  'lib1:build:production': ['lib3:build:ci'],
  'lib3:build:ci': [],
  'lib3:build:production': [],
}
```
Obviously we should avoid the scenario where `lib3:build` is invoked twice with different configurations. I've considered inheriting configuration from the root target and ignore all other occurrences or throwing an error in case such ambiguity is identified, but neither of those seem like a way to go.

So would you consider this functionality as something that is worth adding? If yes, what would be an expected behavior in the scenario described above?

**Comment by dmitry-stepanenko** [comment]
created a draft PR in my fork with partial changes https://github.com/dmitry-stepanenko/nx/pull/1

**Comment by rlmestre** [comment]
`dependsOn` has an expanded syntax ([docs](https://nx.dev/reference/project-configuration#dependson)) that covers configuration forwarding and other params:
```json5
"build": {
   // forward params passed to this target to the dependency targets
  "dependsOn": [{ "projects": "dependencies", "target": "build", "params": "forward" }]
},
```

To run different executors based on config, you could keep them on separate targets and write another executor that handles that logic. It seems to me that being able to depend on a specific configuration implies maintaining multiple targets for each one anyway.

**Comment by dmitry-stepanenko** [comment]
@rlmestre please take a look at the example that I have in the issue description. It doesn't make much sense to maintain intermediate target `build-cloudflare`, which has an only purpose of changing the configuration `build` target is supposed to run with.

**Comment by robwilkes** [comment]
Hmmm, I would like this as well, I have a build target for `google-cloud-build` and would **always** depend on `build:production`. It's a bit of a pain needing to pass `--configuration=production` to `google-cloud-build` every time I run that target and it risks somebody running it in future and forgetting to pass this flag. That said CI/CD will eventually make it a non-issue but in the meantime it seems like a step that could be improved.

**Comment by dmitry-stepanenko** [comment]
@AgentEnder @FrozenPandaz I'm thinking that it would be better to control whether target can be run multiple times on a target itself. I've updated the ticket description with the following:

> 
> As an alternative instead of expecting a` flag that applies to all targets it might be better to have a new `acceptSameTargetVariations` option on the target configuration that we are fine to run multiple times
> ```json
>       "do-stuff": {
>         "executor": "my-executor",
>         "acceptSameTargetVariations": true,
>         "options": {
>           "configFile": "apps/myapp/config.ts"
>         }
>       }
> ```
> 
> And in the same manner as in the approach above we can throw a validation error if `"acceptSameTargetVariations": true` is not present
> ```bash
> >  NX   Could not execute command because the task graph has targets included more than once with different configurations
> 
>   Targets:
>   lib3:build:ci, lib3:build:production
>    
>    If this is intended, set "acceptSameTargetVariations" option to true in target's configuration.
> ```

Let me know if this works better for you, I'll update my PR

**Comment by tnrich** [comment]
@dmitry-stepanenko thanks for taking this issue on. I'd love to see this solved in a clean and simple way. I was surprised when my 
```
"dependsOn": ["build:demo"],
```
didn't just work out of the box. It definitely seems like the right approach to allow this sort of syntax.   

@AgentEnder @FrozenPandaz can we please get this PR in? Thank you!!

**Comment by dimitriy-k** [comment]
I need it too!

`"dependsOn": ["build:demo"],`

**Comment by lppedd** [comment]
Extremely useful, and got here just two days after using Nx, so this should tell you something.  
Bumping for visibility.

**Comment by masonmark** [comment]
Yeah, we've spent several man-hours working around this issue since we adopted Nx some years ago... it comes up on most projects that aren't simple libs, and our engineers have worked around it a dozen different ways by now, and our project.json files are littered with those different workarounds. 

And like other commenters, pretty much everybody asked how to do this within the first week or two of being introduced to Nx. So I'd also like to keep this issue alive for the holidays, too.... 🎅

**Comment by mattfysh** [comment]
Can anyone from nx speak to the resistance against implementing this feature? Is there something we're doing wrong by mixing configuration names together in the task graph, eager to learn if there is a better approach that we should be taking here

**Comment by bsafwen** [comment]
my team decided to use Nx and we stumbled upon this issue.

**Comment by ciriousjoker** [comment]
If this was implemented, here's our usecase:
We have an Angular prerender target, but Angular doesn't support localization with prerendering. This forces us to use one configuration of that prerender target for each language.

We'd like to do this:
```ts
"dependsOn": [
  "prerender:production-en",
  "prerender:production-de",
]
```

**Comment by elbazon** [comment]
Would love to see this feature as-well.
Very usefull.

**Comment by gabrielbryk** [comment]
> Can anyone from nx speak to the resistance against implementing this feature? Is there something we're doing wrong by mixing configuration names together in the task graph, eager to learn if there is a better approach that we should be taking here



> Can anyone from nx speak to the resistance against implementing this feature? Is there something we're doing wrong by mixing configuration names together in the task graph, eager to learn if there is a better approach that we should be taking here

This would be great. This seems like a huge oversight by not supporting this, all my builds are defaulting to their dev configuration and it's forcing me to do some weird workarounds, why not support this?

**Comment by ntinkler-calendly** [comment]
I'll add that this is a giant pain for us.

We make (read - made the mistake of) fairly extensive use of configurations, and right now it's incredibly painful to not be able to pick a configuration as a dependency. 

The existing solutions are hacky - so much so that it's often easier (and frankly more sane) to just bump out to shell scripts with run-command and run the exact commands (and their specific configurations) we need,  but this removes 95% of the utility of NX.

To be extra clear - we don't even fall into the edge cases presented here.  We do not trigger runs that would attempt to build multiple configurations of a target.  But I need the ability to specify a target and a configuration by name in the deps array.  The configuration names are not always the same across targets either, so forwarding args is nowhere close to getting us where we'd like to be.

---

After stewing on this pain for a year+... My stance is now this: 

I don't think configurations should exist if you can't specify them as dependencies.  What's the point of being able to partially override my target config if I can't reliably do it in all places that need it...?  

It seems like it would have been far better to just force us to copy the entire target, instead of pretending this case is supported.  Sadly - this makes the configuration of projects much more verbose and much more prone to code-rot and copy-paste errors.

**Comment by lppedd** [comment]
Eight months after my initial comment on this issue, and my Nx configuration files have become a total mess, I have to spend hours figuring out which duplicated targets to change each time.

This is making me want to get rid of Nx and switch to setting up Gradle tasks, since I'm already using it for Kotlin projects.  
And if I'm at the point of considering Gradle... something enormously wrong is happening.

**Comment by Lp-Francois** [comment]
Hey! We do have the same issue where most of our docker commands need to depends on `build:production`, but we need to slightly change some params like `load: true`, `push: true` and the tags.

Right now it makes everything super verbose and duplicated :( 

We tried to improve this with moving everything we can in `nx.json` as a set of default options for docker targets, that makes it slightly less verbose.

**Comment by hinogi** [comment]
> @dmitry-stepanenko thanks for taking this issue on. I'd love to see this solved in a clean and simple way. I was surprised when my
> 
> ```
> "dependsOn": ["build:demo"],
> ```
> 
> didn't just work out of the box. It definitely seems like the right approach to allow this sort of syntax.
> 
> @AgentEnder @FrozenPandaz can we please get this PR in? Thank you!!
> 
> Here's what I'm hoping to have work in the future: <img alt="image" width="521" src="https://private-user-images.githubusercontent.com/2730609/265833428-f0906a9a-880d-41a3-8540-56aa088b6af8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA2MjY2NTcsIm5iZiI6MTcyMDYyNjM1NywicGF0aCI6Ii8yNzMwNjA5LzI2NTgzMzQyOC1mMDkwNmE5YS04ODBkLTQxYTMtODU0MC01NmFhMDg4YjZhZjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTBUMTU0NTU3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDQwYTYzOTE2MDhmZjYxZjVmNDg1NmQ5MjJhYWNiZWJlOTM2ZTM2ODk0ZmY4N2FmMWQ5OTU4ZjdhNjM5NmI0MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.VqUNN4RKXmDEWtcIugrM35zgIoPNvmgeZMzJPdtWxWE">

I mean, you could make the dirty trick like

```json
{
  "build": {
    ...
    "configuration": {
      ...
    }.
  "build:demo": {
    ...
    "configuration": {
      ...
    }.
}
```

**Comment by AgentEnder** [comment]
Hey - its been a long time since we talked about this as a core team but @FrozenPandaz and I recently sat some time aside to discuss it.

### TLDR
We are going to close the linked PR for now. Its been open too long and we don't want to further string @dmitry-stepanenko along while we try to figure out what the correct path forward is here. His effort is greatly appreciated. We know this is going to be a bit controversial, but we are trying to make the best Nx Nx can be and believe that there is more thought necessary here.

### Why are we closing the linked PR?
While we know as a team that this is a highly requested feature and we appreciate community contributions the PR has been open for too long without response from us. This is not something we take pride in. That said, combined with the fact that significant changes would be necessary and those changes are hard to express as there are still a ton of unknowns, we believe it would be better for this change to come from someone on the core team after a significant amount of design time has been spent thinking throw the problem and edge cases.

### Where do we go from here?
Ok, so if we still think this is a good feature and something we want to have we need to think through some edge cases while also thinking through the initial problem statement. I'm sure everyone here has a valid use case for this feature given the current state of Nx, but maybe there could even be a different change that fits the use case better.

As an example, if targets could have an `extends` property to extend another target on the same project, that may be enough to satisfy some use cases while keeping things distinct targets. 

But, lets continue down the rabbit hole of allowing configurations to be depended on directly rather than another larger change. This sounds simple, and at the base level implementation isn't obscene, but there are some concerns and thus edge cases.

A few of our chief concerns with just allowing this is as follows:
- What happens if multiple instances of the same target end up in a single task graph?
   - Run both? What if they output to the same files, that could easily end up in a state where an unexpected version of artifacts is present when a parent task is executed.
   - Error? What's the suggestion to fix this for an end user? We believe it would be to break things out into separate targets... which is akin to saying "don't use the new feature"
- How should configurations cascade down the task graph?
- 

I'm including a few edge cases below as a sample of what we are thinking about.

#### Localization based configurations

Imagine you have 2 projects, "french-site" and "english-site" that depend on a buildable shared library. That buildable shared library has a target, build, with two configurations ("en, and fr"). The french site's build target has depends on configured as: `[{ target: build, dependencies: true, configuration: fr}]`. The english site has a similar configuration.

```mermaid
graph TD
    B[french-site] -->|"build (fr)"| A[Shared Library]
     C[english-site] -->|"build (en)"| A[Shared Library]
```

The user runs `nx run-many -t build`. What tasks should Nx run, and in what order? The expanded task graph would look something like this naively:

```mermaid
graph TD
    B[french-site:build] --> A["shared-library:(build:fr)"]
    C[english-site:build] --> D["shared-library:(build:en)"]
    E["shared-library:build"]
```

Immediately there is likely to be a problem. Nx will say "I can run shared-library:build, shared-library:build:fr, and shared-library:build:en in parallel, as they all need to be ran and have no dependencies". The outputs of the targets will likely all be written on top of each other leaving the filesystem in a nondeterministic state that isn't likely to produce the desired outputs when the french and english sites build. If things are ran in serial, its even worse as the 3 tasks with no deps will run first, leaving whichever ran last's artifacts to be consumed by both the french and english site.

So we try to avoid that. We can either:
- Tell people that these should be separate targets (build-fr and build-en maybe?)
- Allow specifying inputs / outputs at the configuration level, which isn't currently possible.

Separate targets makes the most sense to us. This is the existing workflow that many in this thread have already adopted, but there is a pain point in that target configurations allow pulling common options to the top level options block rather than duplicating things. This is where we'd potentially propose the addition of an "extends" property inside of a target declaration.

#### Linear Dependency Graph

Imagine you have 3 projects, a b and c. Each have a build target with dev and prod configurations. A depends on B and B on C. A's build target has a dependsOn block defined like: `[{dependencies: true, target: build, configuration: prod}]`. B's dependsOn configuration doesn't specify a specific configuration.

```mermaid
graph TD
    A -->|"build (prod)"| B
     B -->|build| C
```

When the user runs `nx run-many -t build`, what should happen? The initial task graph after some deduplication would look something like this:

```mermaid
graph TD
 A[a:build] --> B[b:build:prod] --> C[c:build:?]
 D[b:build] --> E[c:build]
```

Semantically, we believe that if the dependency between B and C doesn't specify a configuration, it should inherit the configuration from the parent task. So, since B was built prod due to the dependency on A, in the first chain at least, C would build with prod as well.

The second chain is arguably a duplicate execution path that wouldn't be needed, but this may not be the case if for whatever reason the target would fail in one configuration but not its default or similar.

### Conclusion
We understand that this decision might not be ideal for everyone, but we believe it is necessary to ensure the best possible outcome for Nx. We deeply value the community's contributions and encourage continued dialogue and collaboration. Our goal is to revisit this feature with a well-thought-out approach that addresses the complexities and edge cases we've identified. Thank you for your understanding and continued support as we strive to improve Nx together.

**Comment by gcko** [comment]
If you need to call a specific configuration rather than the default, you can work around this issue today by flattening some of the configurations you need into their own distinct task runner. TO BE CLEAR this is absolutely not ideal, but in some cases this may solve your needs.

Example:

You can take the following `test` configuration:
```json
{
"test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "defaultConfiguration": "development",
      "options": {
        "jestConfig": "apps/demo/jest.config.ts"
      },
      "configurations": {
        "development": {
          "watchAll": false,
          "coverageReporters": [
            "text-summary"
          ]
        },
        "pre-commit": {
          "ci": false,
          "bail": 1,
          "watchAll": false,
          "codeCoverage": false
        },
        "ci": {
          "testResultsProcessor": "jest-junit",
          "updateSnapshot": true,
          "ci": true,
          "watchAll": false,
          "codeCoverage": true
        }
      }
    },
}
```

And flatten it into multiple tasks:

```json
...
"test-pre-commit": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "apps/ts/jest.config.ts",
        "ci": false,
        "bail": 1,
        "watchAll": false,
        "codeCoverage": false
      }
    },
    "test-ci": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "testResultsProcessor": "jest-junit",
        "updateSnapshot": true,
        "ci": true,
        "watchAll": false,
        "codeCoverage": true
      }
    },
...
```

Again, this only addresses certain use-cases, but it may be helpful for some of you.

**Comment by TomasChmelik** [comment]
My problem would be solved by having some sort of "private" targets. Targets which don't show up in the Nx Console and can't be started directly from CLI. They can only be started by `dependsOn` of other target.

Then I could run any target I want in that target like:
```json
{
  "targets": {
    "__dont-use__": {
      "command": "nx run project:target:configuration"
    },
    "build": {
      "dependsOn": [
        "__dont-use__"
      ]
    }
  }
}
```

I just don't want to clutter the console with `__dont-use__` target

**Comment by electrocnic** [comment]
What is the status of this?

I would also like to chime in and propose my intuition that led me to this issue:

I would like to be able to add different `dependsOn` for each configuration within a target.
I.e. not globally per target like
```json
{
  "build": {
    ...
    "configurations": {
      "debug": {
         ...
      },
      "prod": {
         ...
      },
    },
    "dependsOn": [...],
  },
  "serve": {
    ...
    "configurations": {
      "debug": {
         ...
      },
      "prod": {
         ...
      },
    },
    "dependsOn": [...],
  }
}

```
but much more like this:
```json
{
  "build": {
    ...
    "configuration": {
      "debug": {
         ...,
         "dependsOn": ["someOtherTarget:debug"],
      },
      "prod": {
         ...,
         "dependsOn": ["someOtherTarget:prod"],
      },
    },
  },
  "serve": {
    ...
    "configuration": {
      "debug": {
         ...,
         "dependsOn": ["someOtherTarget:debug"],
      },
      "prod": {
         ...,
         "dependsOn": ["someOtherTarget:prod"],
      },
    },
  }
}

```

This would be the most intuitive thing I would have done automatically, maybe it could even help with design decisions as I read the previous problem statements.
Maybe this was discussed in your design process anyways.

**Comment by StiliyanKushev** [comment]
+1 Bump, this is a must have feature IMHO.

**Comment by tlaurie-cl** [comment]
In my case I have dev/prod task depending on dev/prod task. Here is what I used based on the other `nx:run-commands` workarounds described in this thread:

```json
"first": {
  "executor": "nx:run-commands",
  "defaultConfiguration": "development",
  "configurations": {
  "development": {
    "command": "echo hello world from first:development"
  },
  "production": {
    "command": "echo hello world from first:production"
  }
},
"first:_development":{
  "executor": "nx:run-commands",
  "options": {
    "command": "npx nx run foo:first:development"
  }
},
"first:_production": {
  "executor": "nx:run-commands",
  "options": {
    "command": "npx nx run foo:first:production"
  }
},
"second": {
  "executor": "nx:run-commands",
  "defaultConfiguration": "development",
  "configurations": {
    "development": {
      "command": "npx nx run foo:_second:_development"
    },
    "production": {
      "command": "npx nx run foo:_second:_production"
    }
  }
},
"_second": {
  "executor": "nx:run-commands",
  "defaultConfiguration": "development",
  "configurations": {
    "development": {
      "command": "echo hello world from second:development"
    },
    "production": {
      "command": "echo hello world from second:production"
    }
  }
},
"_second:_development":{
  "dependsOn": ["first:_development"],
  "executor": "nx:run-commands",
  "options": {
    "command": "npx nx run foo:_second:development"
  }
},
"_second:_production": {
  "dependsOn": ["first:_production"],
  "executor": "nx:run-commands",
  "options": {
    "command": "npx nx run foo:_second:production"
  }
}
```

- It enables `"dependsOn"` to specify configurations by splitting the `first` configurations into separate proxy tasks, `first:_development` and `first:_production`, and then references them with `nx:run-commands`.
- It also enables `"configurations"` to specify `"dependsOn"` by proxying the `second` target and moving it's `"dependsOn"` to it's configuration proxies `second_development` and `second:_production`.

This approach is transparent, doesn't duplicate configuration, and can be used in both directions. However it is ugly, verbose, and places the `"dependsOn"` property away from it's owning target.

---

The cases @AgentEnder described read a lot like the diamond problem. I don't think this is something Nx should be expected to solve, nor something it should gatekeep other developers from solving in their projects.

If `shared-library:build:en` conflicted with `shared-library:build:fr`, *of course* I wouldn't want to run it at the same time.
If I knew that they didn't conflict, *of course* I would accept the task running twice in the same graph.

This is something everyone already had to consider with other build tools or bespoke node scripts. But because of these limitations, devs are forced to also workaround the Nx architecture itself. They can either:

- Replace target `"dependsOn"` with proxy targets and `"nx:run-commands"`, like I've chosen.
- Replace target `"configurations"` with duplicated targets all the way down the task graph.

Neither of these workarounds remove the underlying problem of conflicting targets. But they do remove Nx's value-adds of the task graph and 'reducing repetitive configuration'.

---

I think any of these options would improve the ergonomics around this issue:
- [Support specifying configurations in `"dependsOn"`](https://github.com/nrwl/nx/issues/15064#issue-1588081887) and [add a `"dependsOn"` property to `"configurations"`](https://github.com/nrwl/nx/issues/15064#issuecomment-2637497613).
  - If extra safety is desired: add another `"configurations"` option (`"allowConcurrentWith": [<target>:<config>, ...]`?) to specify what other configurations can be run at the same time. Then if two of the same task end up in the graph that don't have this option, throw an error.
- [Add an "extends" property to enable composable targets](https://github.com/nrwl/nx/issues/15064#issuecomment-2248947857).
- Support a dynamic `project.js` config to enable composable targets.

**Comment by Sanaruca** [comment]
This would also be useful to load environment variables according to the target configuration. 

For example (my case):

Suppose we have `foo` and `bar` projects that could use docker. 

The `foo` service exposes an api that will be used by `bar` service via http, so the `bar` service would make requests to `http://foo:3000`.

Since thanks to nx we can [define environment variables](https://nx.dev/recipes/tips-n-tricks/define-environment-variables) through `.env` files, we could have different .env files for each `bar` target and configuration:
```env
# bar/.build.local.env
FOO_URL=“http://localhost:3000”
```
and
```env
# bar/.build.docker.env
FOO_URL=“http://foo:3000”
````

In each project we have a `docker-build` target to build a docker image that depends on the application build:

Example:
```JSON
{
  "docker-build": {
    "dependsOn": ["build"],
    "command": "docker build ..."
  }
}
```

In this case, if we were to run`nx docker-build bar` this would build the application with the default configuration:

```
$ nx docker-build bar

NX Running target docker-build for project bar and 1 task it depends on:
-------------------------------------------------------------------------
> nx run bar:build:production
...
````
Where `nx run bar:build:production`does load the environment variables from the `.build.local.env` file.

But being able to do something like:

```JSON
{
  "targets": {
    "build": {
      "configurations": {
        "docker": {}
      }
    },
    "docker-build": {
      "dependsOn": ["build:docker"], // <--
      "command": "docker build -f bar/Dockerfile . -t bar"
    }
  }
}
```
Would solve the problem in this specific case or at least that is the result I would expect.

**Comment by phazei** [comment]
@AgentEnder   does it really take 2 years to not get anything done on what should be an incredibly simple feature?

**Comment by austinw-fineart** [comment]
> [@AgentEnder](https://github.com/AgentEnder) does it really take 2 years to not get anything done on what should be an incredibly simple feature?

I'm pretty sure the Nx team doesn't actually use assignees to denote people who are working on an issue, contrary to what most people would expect. Then again, I'm not sure what else you'd use assignees for really.

**Comment by lppedd** [comment]
Here I am again 👀 

For the Nx folks: a "straightforward" (and I put it in quotes because it isn't, but compared to rearchitecting the whole runner...) way to solve this problem would be to offer the possibility to author `project.json` files in TS/JS, e.g. with a `project.nx.(j|t)s`.

I know that there has been some effort into "forcing" the configuration into `package.json`, but, `project.json` is still supported and I believe it will continue to be so.

Allowing authoring configurations in such a way would allow creating multiple targets instead of multiple configurations, **WITHOUT** code duplication, as we can extract and reuse typed plain JS objects.

**Comment by phazei** [comment]
Still relevant

**Comment by bojanbass** [comment]
Surely relevant and would help a lot in flexibility.

**Comment by peshala-prabhapoorna** [comment]
Still relevant and would be quite useful. Alternative is to create dedicated target with the desired config, but the config override is much more elegant.
